### PR TITLE
Issue156

### LINKED ISSUE #156: Battery level dependencies/prereqs
We want to create lists of batteries that should have been done and lists of batteries that should have not been done to determine whether a work is eligible for a particular battery. Choosing these battery prerequisites should be on the battery creation screen.

This can either be taken care of in the same area as PC/chrome checks, or potentially with qualifications.

@rwblair.

### PR DESCRIPTION
**Pull request by rwblair**
Commits to resolve #156. Works by generating a list of completed batteries a worker has done based on comparing a list of experiments completed for an assignment to the list of experiments for a given battery that is a dependency.  

This check is done when serving a hit to a turker. I also added the check to when a local battery is being deployed in expdj/apps/experiments/views.py, not sure if its useful there, can remove if not.

New error page detailing missing batteries added.

Updated battery edit form to use select2 to select required batteries.

### CODE REVIEW

File: expdj/apps/experiments/models.py [hunk 1]
```diff
@@ -176,6 +202,61 @@ def save(self, *args, **kwargs):
...
+        '''
+        check_battery_dependencies looks up all of a workers completed 
+        experiments in a result object and places them in a dictionary 
+        organized by assignment_id. Each of these buckets of results is
+        iterated through to check that every experiment in a battery has
+        been completed. In this way a list of batteries that a worker has 
+        completed is built. This list is then compared to the lists of 
+        required and restriected batteries to determine if the worker is 
+        elidgble to attempt the current(self) battery.
+        '''
```
Comment (vsoch) [inline expdj/apps/experiments/models.py]: eligible :)

File: expdj/apps/experiments/models.py [hunk 2]
```diff
@@ -176,6 +202,61 @@ def save(self, *args, **kwargs):
...
+        worker_results = turk.models.Result.objects.filter(
```
Comment (vsoch) [inline expdj/apps/experiments/models.py]: I like this function, but I don't think it should be part of the Battery Model. It integrates functionality from the Turk and Experiments module, and seems more fit for a tasks.py function. Depending on the user of the experiment factory, a battery does not necessarily need to have an understanding of a worker status, and I'd go as far to say that an ideal application would have subject management separately from the experiments. I think this should be moved to tasks.py.

File: expdj/apps/turk/views.py [hunk 3]
```diff
@@ -127,6 +127,13 @@ def serve_hit(request,hid):
         # Get Experiment Factory objects for each
         worker = get_worker(aws["worker_id"])
 
+        missing_batteries, blocking_batteries = battery.check_battery_dependencies(aws["worker_id"])
```
Comment (vsoch) [inline expdj/apps/turk/views.py]: if this is repeated twice, shouldn't it be a function?

File: expdj/apps/turk/views.py [hunk 4]
```diff
@@ -208,6 +215,15 @@ def preview_hit(request,hid):
         hit =  get_hit(hid,request)
         battery = hit.battery
         context = get_amazon_variables(request)
+
+        missing_batteries, blocking_batteries = battery.check_battery_dependencies(context["worker_id"])
```
Comment (vsoch) [inline expdj/apps/turk/views.py]: Do we want this in the preview hit function?

File: expdj/apps/experiments/models.py [hunk 5]
```diff
@@ -210,8 +210,8 @@ def check_battery_dependencies(self, worker_id):
         iterated through to check that every experiment in a battery has
         been completed. In this way a list of batteries that a worker has 
         completed is built. This list is then compared to the lists of 
-        required and restriected batteries to determine if the worker is 
-        elidgble to attempt the current(self) battery.
+        required and restricted batteries to determine if the worker is 
+        eligible to attempt the current(self) battery.
         '''
```
Comment (vsoch) [inline expdj/apps/experiments/models.py]: +1 :)

File: expdj/apps/experiments/models.py (hunk 2 above)
Comment (rwblair) [inline expdj/apps/experiments/models.py]: It won't be used as an asynchronous task, but it looks like there are similar supporting functions there. Moved function to tasks.py in commit 8ffabfa.

File: expdj/apps/turk/views.py (hunk 3 above)
Comment (rwblair) [inline expdj/apps/turk/views.py]: Wrapped redundant code in a new function check_battery view in commit 8ffabfa

File: expdj/apps/turk/views.py (hunk 4 above)
Comment (rwblair) [inline expdj/apps/turk/views.py]: I think it would help discourage people from accepting a hit that we don't want them to do.

**Comment by vsoch** [comment]
We can deploy this to test when @IanEisenberg is ready. I wouldn't do this if this were a real deployed application, but only our lab is using it, so it doesn't matter if it bugs out and we fix/revert it.

**Comment by rwblair** [comment]
@IanEisenberg, @jkl071 and myself have tested these changes on the testing server and they are now working as expected.

I haven't figured out a way to get access to a users worker ID in an experiment preview, so a HIT must be accepted and go through the advertisement/consent form before these dependency checks can be made.

I tried to have a return hit link on the dependencies not met page, but haven't been able to construct a valid url that mturk.com accepts for returning a hit. The hold up here is that the hit ID listed in "mturk.com/mturk/accept" and the hit ID in the link to "mturk.com/mturk/return" on the accept page are different. The ID in the return URL is not one that I've been able to find in the database. There might be a way to generate this new hit ID with information we have on hand, but I haven't been able to find any documentation for it.

So the current state of the PR; contains a basic page preventing workers from continuing a hit after they have accepted it if they have not met the requirements or have met the restrictions on a battery, and alerts them about the reason that they cannot continue.

**Pull request closed without merging**